### PR TITLE
New IronSource Adapter To Support Banner

### PR DESCRIPTION
**Pull request by alon357**
New IronSource Adapter:
* Contains new Adapter to support IronSource Banner.
* Update IronSource Interstitial and Rewarded API that was deprecated.

### CODE REVIEW

**Comment by joshh-devrel** [comment]
@alon357 Can you rebase your pull request and resolve the conflicts? Thanks!

File: ThirdPartyAdapters/ironsource/ironsource/libs/PLACE_IRONSOURCE_SDK_JAR_HERE.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+The jar file for the IronSource Android SDK should be placed in this directory and called 'ironsource' .
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/libs/PLACE_IRONSOURCE_SDK_JAR_HERE.txt:1]: This isn't needed as publishers can define where to put the IronSouce Android SDK file.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 2]
```diff
@@ -4,200 +4,375 @@
...
-    Log.e(TAG, showError.getMessage());
-  }
-  // endregion
+public class IronSourceAdapter implements MediationInterstitialAdapter, MediationBannerAdapter, IronSourceAdapterListener {
+
+    /**
+     * Mediation interstitial ad listener used to forward interstitial events from IronSource SDK to
+     * Google Mobile Ads SDK.
+     */
+    private MediationInterstitialListener mInterstitialListener;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java:45]: Can you remove the `m` prefix?

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 3]
```diff
@@ -4,200 +4,375 @@
...
+
+    /**
+     * The view for the banner instance.
+     */
+    private ISDemandOnlyBannerLayout mIronSourceBannerLayout;
+
+    /**
+     * Callback object for Google's Banner Lifecycle.
+     */
+    private MediationBannerListener mediationBannerListener;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java:60]: Can update the variable name to `bannerListener` to be consistent to the one for interstitial?

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 4]
```diff
@@ -4,200 +4,375 @@
...
+
+    /**
+     * This is the id of the instance to be shown.
+     */
+    private String mInstanceID;
+
+    /**
+     * The view for the banner instance.
+     */
+    private ISDemandOnlyBannerLayout mIronSourceBannerLayout;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java:55]: Can you also remove `m` prefixes for these variables?

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 5]
```diff
@@ -4,200 +4,375 @@
...
+    }
+
+    @Override
+    public void onPause() {
+
+    }
+
+    @Override
+    public void onResume() {
+    }
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java:110]: Can you remind me again when will these callbacks be invoked?

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java (hunk 3 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java]: Done Thanks

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java (hunk 4 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java]: Done Thanks

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java (hunk 2 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java]: Done Thanks

**Comment by joshh-devrel** [comment]
@alon357 
Can you rebase your branch? I am still seeing `FacebookAdapter.java` in the diff.
Also, please sign the CLA.

File: ThirdPartyAdapters/ironsource/ironsource/build.gradle [hunk 6]
```diff
@@ -8,7 +8,9 @@ ext {
     // String property to store the proper name of the mediation network adapter.
     adapterName = "IronSource"
     // String property to store version name.
-    stringVersion = "7.2.3.1.0"
+
+    stringVersion = "7.2.4.0"
+
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/build.gradle:13]: Nit, remove extra lines.

File: ThirdPartyAdapters/ironsource/ironsource/build.gradle [hunk 7]
```diff
@@ -31,7 +33,9 @@ android {
 }
 
 dependencies {
-    implementation 'com.ironsource.sdk:mediationsdk:7.2.3.1'
+
+    implementation 'com.ironsource.sdk:mediationsdk:7.2.4'
+
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/build.gradle:38]: Nit, remove extra lines.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java [hunk 8]
```diff
@@ -50,4 +64,38 @@ static synchronized void sendEventOnUIThread(Runnable runnable) {
...
+  @Nullable
+  public static ISBannerSize getISBannerSize(@NonNull Context context,
+                                             @NonNull AdSize adSize) {
+    ArrayList<AdSize> potentials = new ArrayList<>();
+    potentials.add(AdSize.BANNER);
+    potentials.add(AdSize.MEDIUM_RECTANGLE);
+    potentials.add(AdSize.LARGE_BANNER);
+
+
+
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java:77]: Nit, remove extra lines.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java [hunk 9]
```diff
@@ -50,4 +64,38 @@ static synchronized void sendEventOnUIThread(Runnable runnable) {
...
+        }
+
+      return new ISBannerSize(closestSize.getWidth(), closestSize.getHeight());
+    }
+
+    return null;
+  }
+
+  @NonNull
+  static final AdError createAdError(int errorCode, @NonNull String description) {
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java:96]: I don't see this method is used anywhere.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java [hunk 10]
```diff
@@ -47,15 +50,21 @@ public class IronSourceMediationAdapter extends Adapter
...
+    /**
+     * Banner size mismatch.
+     */
+    public static final int ERROR_BANNER_SIZE_MISMATCH = 101;
+
   /**
    * Server parameters (e.g. placement ID) are nil.
    */
-  public static final int ERROR_INVALID_SERVER_PARAMETERS = 101;
+  public static final int ERROR_INVALID_SERVER_PARAMETERS = 110;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java:67]: Revert back to `101`.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java [hunk 11]
```diff
@@ -47,15 +50,21 @@ public class IronSourceMediationAdapter extends Adapter
...
+          ERROR_BANNER_SIZE_MISMATCH,
       })
   public @interface AdapterError {
 
   }
 
+    /**
+     * Banner size mismatch.
+     */
+    public static final int ERROR_BANNER_SIZE_MISMATCH = 101;
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java:62]: Update to `105`.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java [hunk 12]
```diff
@@ -50,4 +64,38 @@ static synchronized void sendEventOnUIThread(Runnable runnable) {
     }
     uiHandler.post(runnable);
   }
+
+  @Nullable
+  public static ISBannerSize getISBannerSize(@NonNull Context context,
```
Comment (joshh-devrel) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java:69]: This is also not used anywhere, this should be called before loading the banner ad.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java (hunk 9 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java]: Removed the function as it is not in use

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java (hunk 10 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java]: Fixed thanks

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java (hunk 11 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java]: Fixed thanks

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java (hunk 12 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapterUtils.java:69]: This is used in our IronSourceAdapter class in the function requestBannerAd (line 328)
Once we create the banner layout(line 359) we set the size

**Comment by alon357** [comment]
Hey @joshh-devrel 
I have removed Facebook and the CLA says I already signed it (Are you referring to the organization CLA?)
Thanks

**Comment by alon357** [comment]
Hey @joshh-devrel 
Any update regarding the PR?

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java [hunk 13]
```diff
@@ -76,14 +85,23 @@ void initIronSourceSDK(@Nullable Context context, @Nullable String appKey,
...
 
     isInitialized.set(true);
     listener.onInitializeSuccess();
   }
 
-  void loadInterstitial(@NonNull String instanceId, @NonNull IronSourceAdapter adapter) {
+  void loadInterstitial(@Nullable Context context,@NonNull String instanceId, @NonNull IronSourceAdapter adapter) {
+    if (!(context instanceof Activity)) {
+      String errorMessage = String
+              .format(ERROR_REQUIRES_ACTIVITY_CONTEXT + "IronSource requires an Activity context to load ads.");
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java:98]: Is this formatted properly?
I feel like this should be something like `("[%d] errorMessage", ERROR_REQUIRES_ACTIVITY_CONTEXT)`, or just without the ERROR CODE and ot should just be a string literal

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java [hunk 14]
```diff
@@ -120,7 +147,28 @@ void loadRewardedVideo(@NonNull String instanceId, @NonNull IronSourceMediationA
...
+    }
+
+    if (TextUtils.isEmpty(instanceId)) {
+      AdError loadError = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
+              "Missing or invalid instance ID.", ERROR_DOMAIN);
+      adapter.onAdFailedToLoad(loadError);
+      return;
+    }
+
+    registerISBannerAdapter(instanceId, new WeakReference(adapter));
```
Comment (joshuajplagonera) [inline ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java:169]: Can you confirm if it is possible to load 2 IronSource banner ads simultaneously using the same Instance ID?

**Review by joshuajplagonera — CHANGES_REQUESTED** [review]
Added comments.

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 15]
```diff
@@ -4,200 +4,374 @@
...
-    Log.e(TAG, showError.getMessage());
-  }
-  // endregion
+public class IronSourceAdapter implements MediationInterstitialAdapter, MediationBannerAdapter, IronSourceAdapterListener {
+
+    /**
+     * Mediation interstitial ad listener used to forward interstitial events from IronSource SDK to
+     * Google Mobile Ads SDK.
+     */
+    private MediationInterstitialListener InterstitialListener;
```
[the remark this answers is not in the document]
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java]: Fixed and rebased

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java [hunk 16]
```diff
@@ -76,14 +85,23 @@ void initIronSourceSDK(@Nullable Context context, @Nullable String appKey,
 
     IronSource.setMediationType(MEDIATION_NAME + ADAPTER_VERSION_NAME);
     Log.d(TAG, "Initializing IronSource SDK with app key: " + appKey);
-    IronSource.initISDemandOnly(activity, appKey, IronSource.AD_UNIT.INTERSTITIAL,
-        IronSource.AD_UNIT.REWARDED_VIDEO);
+    IronSource.initISDemandOnly(context, appKey, IronSource.AD_UNIT.INTERSTITIAL,
```
[the remark this answers is not in the document]
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java:88]: IronSource require Activity Context and wont work with application Context

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java (hunk 13 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java]: Fixed

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java (hunk 14 above)
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceManager.java:169]: Loading the same instance twice will create a scenario of the second load will override the first load

File: ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java [hunk 17]
```diff
@@ -4,200 +4,374 @@
...
+    public void requestBannerAd(@NonNull Context context,
+                                @NonNull MediationBannerListener listener,
+                                @NonNull Bundle serverParameters, @NonNull AdSize adSize,
+                                @NonNull MediationAdRequest mediationAdRequest,
+                                @Nullable Bundle bundle1) {
+        ISBannerSize bannerSize = IronSourceAdapterUtils.getISBannerSize(context, adSize);
+
+        if (!(context instanceof Activity)) {
+            AdError errorMessage = new AdError(ERROR_REQUIRES_ACTIVITY_CONTEXT, "IronSource requires an Activity context to load ads." +
+                    "adSize",
```
[the remark this answers is not in the document]
Comment (alon357) [reply ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceAdapter.java:331]: no as IronSource require context of activity and fail if passing application context

**Comment by joshh-devrel** [comment]
@alon357 It doesn't seem like you have rebased successfully,  there are still 69 commits and 83 files changed.
image

**Comment by alon357** [comment]
Hey @joshh-devrel 
There is something weird with the rebase.
It tells me there is nothing to rebase and if I check the files that show as different from the main branch it is identical.(Although it shows diff in git)
Any advise here?

**Pull request closed without merging**